File: client/Scripts/ClientAPI.py
#
#
#  The Multiverse Platform is made available under the MIT License.
#
#  Copyright (c) 2012 The Multiverse Foundation
#
#  Permission is hereby granted, free of charge, to any person 
#  obtaining a copy of this software and associated documentation 
#  files (the "Software"), to deal in the Software without restriction, 
#  including without limitation the rights to use, copy, modify, 
#  merge, publish, distribute, sublicense, and/or sell copies 
#  of the Software, and to permit persons to whom the Software 
#  is furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be 
#  included in all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
#  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES 
#  OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
#  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
#  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
#  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE 
#  OR OTHER DEALINGS IN THE SOFTWARE.
#
#  

import Multiverse.Config
import Multiverse.Base.ClientAPI
import Multiverse.Base
import Multiverse.Gui
import Multiverse.Interface
import Multiverse.Voice
import Axiom.Core
import Axiom.MathLib
from Axiom.MathLib import Quaternion, Vector3, Vector4
from Axiom.Core import ColorEx
import SceneNode
import AnimationState
import WorldObject
import Model
import ParticleSystem
import Decal
import SoundSource
import Animation
import NodeAnimationTrack
import NodeKeyFrame
import PropertyAnimationTrack
import PropertyKeyFrame
import MorphAnimationTrack
import MorphKeyFrame
import Light
import Camera
import Compositor
import Input
import Interface
import SystemStatus
import Voice
import Multiverse.Lib.TextureFetcher

from log4net import LogManager

from Axiom.Graphics import SceneBlendFactor as SceneBlendFactor
from Axiom.Graphics import CompareFunction as CompareFunction
from Axiom.Graphics import CullingMode as CullingMode
from Axiom.Graphics import ManualCullingMode as SoftwareCullingMode
from Axiom.Graphics import SceneDetailLevel as SceneDetailLevel
from Axiom.Graphics import Shading as Shading
from Axiom.Graphics import LayerBlendOperation as LayerBlendOperation
from Axiom.Graphics import LayerBlendOperationEx as LayerBlendOperationEx
from Axiom.Graphics import LayerBlendSource as LayerBlendSource
from Axiom.Graphics import FilterOptions as FilterOptions
from Axiom.Graphics import TextureAddressing as TextureAddressing
from Axiom.Graphics import TextureType as TextureType
from Axiom.Media import PixelFormat as PixelFormat

import GPUProgramType
import Material
import Technique
import Pass
import TextureUnit
import EditableImage
from Axiom.SceneManagers.Multiverse import ShadowTechnique as ShadowTechnique
from Axiom.Graphics import MaterialManager as _MaterialManager
from Axiom.SceneManagers.Multiverse import DefaultLODSpec
from Multiverse.Base import ScriptableLODSpec

from HardwareCaps import HardwareCaps as _HardwareCaps
from World import World as _World
from Network import Network as _Network

import System.Diagnostics

_client = Multiverse.Base.Client.Instance
_worldManager = _client.WorldManager
_sceneManager = _worldManager.SceneManager
_debug = _client.doTraceConsole
_log = Multiverse.Base.ClientAPI.Log

UnitsPerMeter = Multiverse.Base.Client.OneMeter
ScreenshotPath = Multiverse.Base.Client.ScreenshotPath

def NumExistingScreenshots():
    return Multiverse.Base.Client.NumExistingScreenshots

def GetNextScreenshotFilename():
    return Multiverse.Base.Client.GetNextScreenshotFilename(ScreenshotPath)

def GetNextScreenshotFilename(screenshotPath, extension):
    return Multiverse.Base.Client.GetNextScreenshotFilename(screenshotPath, extension)

def Log(msg):
    LogInfo(msg)

def DebugLog(msg):
    LogDebug(msg)

def LogDebug(msg):
    _log.Debug(msg)
            
def LogInfo(msg):
    _log.Info(msg)
            
def LogWarn(msg):
    _log.Warn(msg)
            
def LogError(msg):
    _log.Error(msg)
            
def Write(msg):
    _log.Debug(msg)
    Interface.DispatchEvent("CHAT_MSG_SYSTEM", [msg, ""])
    
def DebugWrite(msg):
    if _debug:
        Write(msg)

try:
  from Multiverse.Movie import IMovie as Movie
  from Multiverse.Movie import ICodec as MovieCodec
  from Multiverse.Movie import Manager as MovieManager
  from Multiverse.Movie import IMovieTexture as MovieTexture
  from Multiverse.Movie import MovieTextureSource as MovieTextureSource
except:
  Log("Exception on movie import, plugin not loaded")
  
from Multiverse.Web import Browser as WebBrowser

def RegisterEffect(effectName, effectFunc):
    _deprecated("1.1", "ClientAPI.RegisterEffect()", "ClientAPI.World.RegisterEffect()")
    World.RegisterEffect(effectName, effectFunc)
    
def InvokeEffect(effectName, oid, args):
    _deprecated("1.1", "ClientAPI.InvokeEffect()", "ClientAPI.World.InvokeEffect()")
    World.InvokeEffect(effectName, oid, args)

def GetLocalOID():
    return _worldManager.GetLocalOid()

def GetPlayerObject():
    if _worldManager.Player != None:
        return WorldObject._GetExistingWorldObject(_worldManager.Player)
    else:
        return None
    
def GetCurrentTime():
    return Multiverse.Base.WorldManager.CurrentTime
    
def GetObjectByOID(oid):
    _deprecated("1.1", "ClientAPI.GetObjectByOID()", "ClientAPI.World.GetObjectByOID()")
    return World.GetObjectByOID(oid)
    
def GetObjectByName(name):
    _deprecated("1.1", "ClientAPI.GetObjectByName()", "ClientAPI.World.GetObjectByName()")
    return World.GetObjectByName(name)
    
def GetObjectNodeNames():
    _deprecated("1.1", "ClientAPI.GetObjectNodeNames()", "ClientAPI.World.WorldObjectNames")
    return World.WorldObjectNames
    
def GetSoundSource(soundFile, position, looping=False, gain=1.0, ambient=False, local=True):
    realSoundSource = Multiverse.Base.SoundManager.Instance.GetSoundSource(soundFile, ambient, local)
    soundSource = SoundSource._ExistingSoundSource(realSoundSource)
    soundSource.Looping = looping
    soundSource.Gain = gain
    if ambient == False:
        soundSource.Position = position
        
    return soundSource
    

def CreatePoseAnimation(meshName, name, totalTime, frameDescList):
    # name is the name of the new animation
    # totalTime is the length of the animation in seconds
    # frameDescList is a list of frame descriptions.
    #
    # Each frame description describes a key frame, and consists of a tuple of 2 elements.
    # The first element is the time offset for the key frame.
    # The second element is a list of 2 element tuples, with the first element being the name of pose,
    # and the second element being the weight for that pose.  All poses not named are assumed to be weight 0.
        
    mesh = Axiom.Core.MeshManager.Instance.Load("facial.mesh");
    anim = mesh.CreateAnimation(name, totalTime)
    track = anim.CreateVertexTrack(4, Axiom.Animating.VertexAnimationType.Pose);
        
    for time, poses in frameDescList:
        keyframe = track.CreateVertexPoseKeyFrame(time)
            
        for poseName, weight in poses:
            keyframe.AddPoseReference(mesh.GetPoseIndex(poseName), weight)
                
    # force update of animation states
    #self._Entity.Mesh = mesh


def CreateMorphAnimation(model, subMeshName, name, totalTime, frameDescList):
    # name is the name of the new animation
    # totalTime is the length of the animation in seconds
    # frameDescList is a list of frame descriptions.
    #
    # Each frame description describes a key frame, and consists of a tuple of 2 elements.
    # The first element is the time offset for the key frame.
    # The second element is a list of 2 element tuples, with the first element being the name of pose,
    # and the second element being the weight for that pose.  All poses not named are assumed to be weight 0.
        
    anim = model.CreateAnimation(name, totalTime)
    track = anim.CreateMorphTrack(subMeshName)
        
    for time, poses in frameDescList:
        keyframe = track.CreateKeyFrame(time)
            
        for poseName, weight in poses:
            keyframe.AddMorphTarget(poseName, weight)
                
    # force update of animation states
    #self._Entity.Mesh = mesh

def LaunchBrowser(url):
    if url.startswith("http:") or url.startswith("https:"):
        System.Diagnostics.Process.Start(url)
    else:
        Write("browser URLs must use http: or https:")

_newWorldEvents = ['WorldInitialized', 'ObjectAdded', 'ObjectRemoved', 'AmbientLightChanged', 'LightAdded', 'LightRemoved']

def RegisterEventHandler(eventName, eventHandler):
    if eventName in _newWorldEvents:
        _deprecated("1.1", "ClientAPI.RegisterEventHandler('" + eventName + "')", "ClientAPI.World.RegisterEventHandler('" + eventName + "')")
        World.RegisterEventHandler(eventName, eventHandler)
    elif eventName == 'WorldConnect':
        Multiverse.Base.ClientAPI.WorldConnect += eventHandler
    elif eventName == 'WorldDisconnect':
        Multiverse.Base.ClientAPI.WorldDisconnect += eventHandler
    elif eventName == 'FrameStarted':
        Multiverse.Base.ClientAPI.FrameStarted += eventHandler
    elif eventName == 'FrameEnded':
        Multiverse.Base.ClientAPI.FrameEnded += eventHandler
    elif eventName == 'TargetChanged':
        LogError("ClientAPI.RegisterEventHandler('TargetChanged') is no longer available")
        _deprecated("1.1", "ClientAPI.RegisterEventHandler('" + eventName + "')", "a subscription to the 'PLAYER_TARGET_CHANGED' UI Event")
    elif eventName == 'PlayerInitialized':
        _worldManager.PlayerInitializedEvent += eventHandler
    else:
        ClientAPI.LogError("Invalid event name '%s' passed to ClientAPI.RegisterEventHandler" % str(eventName))

        
def RemoveEventHandler(eventName, eventHandler):
    if eventName in _newWorldEvents:
        _deprecated("1.1", "ClientAPI.RemoveEventHandler('" + eventName + "')", "ClientAPI.World.RemoveEventHandler('" + eventName + "')")
        World.RemoveEventHandler(eventName, eventHandler)
    elif eventName == 'WorldConnect':
        Multiverse.Base.ClientAPI.WorldConnect -= eventHandler
    elif eventName == 'WorldDisconnect':
        Multiverse.Base.ClientAPI.WorldDisconnect -= eventHandler
    elif eventName == 'FrameStarted':
        Multiverse.Base.ClientAPI.FrameStarted -= eventHandler
    elif eventName == 'FrameEnded':
        Multiverse.Base.ClientAPI.FrameEnded -= eventHandler
    elif eventName == 'TargetChanged':
        LogError("ClientAPI.RemoveEventHandler('TargetChanged') is no longer available")
        _deprecated("1.1", "ClientAPI.RemoveEventHandler('" + eventName + "')", "a subscription to the 'PLAYER_TARGET_CHANGED' UI Event")
    elif eventName == 'PlayerInitialized':
        _worldManager.PlayerInitializedEvent -= eventHandler
    else:
        ClientAPI.LogError("Invalid event name '%s' passed to ClientAPI.RemoveEventHandler" % str(eventName))
        
def _WorldInitHandler(sender, args):
    global ShadowConfig
    global OceanConfig
    global FogConfig
    global AmbientLight
    global PlayerCamera
    ShadowConfig = _sceneManager.ShadowConfig
    OceanConfig = _sceneManager.OceanConfig
    FogConfig = _sceneManager.FogConfig
    AmbientLight = _sceneManager.AmbientLightConfig

def SetSkyBox(materialName, enable = True):
    _deprecated("1.1", "ClientAPI.SetSkyBox()", "ClientAPI.World.SetSkyBox()")
    World.SetSkyBox(materialName, enable)
    
def GetTerrainHeight(x, z):
    _deprecated("1.1", "ClientAPI.GetTerrainHeight()", "ClientAPI.World.GetTerrainHeight()")
    return World.GetTerrainHeight(x,z)
    
def GetTerrainHeightVector(v):
    _deprecated("1.1", "ClientAPI.GetTerrainHeightVector()", "ClientAPI.World.GetTerrainHeightVector()")
    return World.GetTerrainHeightVector(v)

def GetPlayerCamera():
    return Camera._ExistingCamera(_client.Camera)
    
def GrabPlayerCamera():
    InputHandler.CameraGrabbed = True
    
def ReleasePlayerCamera():
    InputHandler.CameraGrabbed = False
    
def GetMousePosition():
    mousePos = Multiverse.Gui.GuiSystem.Instance.MousePosition
    return mousePos.X, mousePos.Y
    
def GetScreenSize():
    _deprecated("1.1", "ClientAPI.GetScreenSize()", "ClientAPI.HardwareCaps.WindowSize")
    return HardwareCaps.WindowSize
    
def PickTerrain(x, y):
    _deprecated("1.1", "ClientAPI.PickTerrain()", "ClientAPI.World.PickTerrain()")
    return World.PickTerrain(x, y)

def SetTerrainDisplay(terrainDisplay):
    _deprecated("1.1", "ClientAPI.SetTerrainDisplay()", "ClientAPI.World.DisplayTerrain")
    World.DisplayTerrain = terrainDisplay    
    
def GetCharacterEntries():
    """Get the list of characters that were provided by the remote server.
       This is a list of objects of type CharacterEntry, each of which
       contains the properties of the character."""
    return _client.CharacterEntries

def Exit():
    """Notify the client that it should shut down.  This sets an internal
       flag so that the client can shut down gracefully as soon as it is
       possible to do so."""
    return _client.RequestShutdown()

def IsWorldLocal():
    """Expose the client flag that indicates whether we are using a locally
       created world.  This allows scripts to tell if we are communicating
       with the world server, or just showing a local scene."""
    _deprecated("1.1", "ClientAPI.IsWorldLocal()", "ClientAPI.World.IsWorldLocal")
    return World.IsWorldLocal
    
def GetMaterial(name):
    mat = _MaterialManager.Instance.GetByName(name)
    if mat is not None:
        return Material._ExistingMaterial(mat)
    return None

def SetTerrainLODSpec(tilesPerPageDelegate, metersPerSampleDelegate, pageSize, visPageRadius):
    """Create a ScriptableLODSpec instance, passing the four arguments to the
       constructor, and set the WorldManager's WorldLODSpec property to the instance.
       Note well: This method must be called in Startup.py so that it takes
       effect before the world geometry is created."""
    scriptableLODSpec = ScriptableLODSpec(tilesPerPageDelegate, metersPerSampleDelegate, pageSize, visPageRadius)
    _worldManager.WorldLODSpec = scriptableLODSpec

def SetTerrainLODSpecPageSize(pageSize):
    """Create a DefaultLODSpec instance, passing the two arguments to the
       constructor, and set the WorldManager's WorldLODSpec property to the instance.
       Note well: This method must be called in Startup.py so that it takes
       effect before the world geometry is created."""
    defaultLODSpec = DefaultLODSpec(pageSize)
    _worldManager.WorldLODSpec = defaultLODSpec

def RegisterObjectPropertyChangeHandler(propName, handler):
    _deprecated("1.1", "ClientAPI.RegisterObjectPropertyChangeHandler()", "ClientAPI.World.RegisterObjectPropertyChangeHandler()")
    World.RegisterObjectPropertyChangeHandler(propName, handler)

def RemoveObjectPropertyChangeHandler(propName, handler):
    _deprecated("1.1", "ClientAPI.RemoveObjectPropertyChangeHandler()", "ClientAPI.World.RemoveObjectPropertyChangeHandler()")
    World.RemoveObjectPropertyChangeHandler(propName, handler)
    
def GetVolumeRegions(location):
    _deprecated("1.1", "ClientAPI.GetVolumeRegions()", "ClientAPI.World.GetVolumeRegions()")
    return World.GetVolumeRegions(location)
    
def GetLight(name):
    _deprecated("1.1", "ClientAPI.GetLight()", "ClientAPI.World.GetLight()")
    return World.GetLight(name)

def SetMaterialScheme(scheme):
    vp = _sceneManager.CurrentViewport
    vp.MaterialScheme = scheme

def GetMaterialScheme():
    vp = _sceneManager.CurrentViewport
    return vp.MaterialScheme
    
def SetTextureNameOverride(name):
    Axiom.Core.TextureManager.Instance.OverrideName = name

def GetLightNames():
    _deprecated("1.1", "ClientAPI.GetLightNames()", "ClientAPI.World.LightNames")
    return World.LightNames

def RegisterCommandHandler(command, hander):
    LogError("ClientAPI.RegisterCommandHandler is no longer available")
    _deprecated("1.1", "ClientAPI.RegisterCommandHandler", "MarsCommand.RegisterCommandHandler")
        
def _deprecated(version, oldMethod, newMethod):
    Multiverse.Base.ClientAPI.ScriptDeprecated(version, oldMethod, newMethod)
    
Multiverse.Base.ClientAPI.WorldInitialized += _WorldInitHandler
RootSceneNode = SceneNode.SceneNode(_sceneManager.RootSceneNode)
HardwareCaps = _HardwareCaps()
World = _World()
Network = _Network()
InputHandler = None

Log("ClientAPI.py loaded")

#
# Newly added methods to remove lower level calls.  These may need to
# be moved into appropriate modules.
#
def GetMouseoverTarget():
    mousePos = Multiverse.Gui.GuiSystem.Instance.MousePosition
    windowSize = HardwareCaps.WindowSize
    if windowSize is None:
        # We're not initialized yet
        Log("WindowSize not initialized")
        return None
    width, height = windowSize
    objNode = _client.CastRay(mousePos.X / width, mousePos.Y / height)
    if objNode is None:
        return None
    else:
        return World.GetObjectByOID(objNode.Oid)

def TakeScreenshot():
    _client.TakeScreenshot()

def ToggleRenderMode():
    _client.ToggleRenderMode()

def ToggleTexture():
    _client.ToggleTexture()

def ToggleBoundingBoxes():
    _client.ToggleBoundingBoxes()

def ToggleRenderCollisionVolumes():
    _client.ToggleRenderCollisionVolumes()

def GetScreenPosition(pos):
    return _client.GetScreenPosition(pos)

def SetClientParameter(name, val):
    return Multiverse.Config.ParameterRegistry.SetParameter(name, val)

def GetClientParameter(name):
    return Multiverse.Config.ParameterRegistry.GetParameter(name)

_rand = System.Random()

def Random(arg):
    return _rand.Next(arg)
    
def RandomFloat(arg):
    return arg * float(_rand.NextDouble())
    
def FetchRemoteTexture(url, textureName, handler, destWidth=0, destHeight=0, keepAspect=False, fillColor=ColorEx.Black, authUser=None, authPW=None, authDomain=None):
    Multiverse.Lib.TextureFetcher.TextureFetcher.Instance.FetchTexture(url, textureName, handler, destWidth, destHeight, keepAspect, fillColor, authUser, authPW, authDomain)

def SetProfileMarker(message, color=ColorEx.Red):
    _sceneManager.SetProfileMarker(color, message)
    
def GetAssetPath(assetName):
    return Axiom.Core.ResourceManager.ResolveCommonResourceData(assetName)

def GetParameter(paramName):
    """This exposes parameters that may have been set on the command line."""
    return _client.GetParameter(paramName)

def HasParameter(paramName):
    return _client.HasParameter(paramName)
    
def SetParameter(paramName, paramValue):
    _client.SetParameter(paramName, paramValue)

File: client/Base/ClientAPI.cs
/********************************************************************

The Multiverse Platform is made available under the MIT License.

Copyright (c) 2012 The Multiverse Foundation

Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation 
files (the "Software"), to deal in the Software without restriction, 
including without limitation the rights to use, copy, modify, 
merge, publish, distribute, sublicense, and/or sell copies 
of the Software, and to permit persons to whom the Software 
is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be 
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE 
OR OTHER DEALINGS IN THE SOFTWARE.

*********************************************************************/

using System;
using System.Collections.Generic;
using System.Text;
using System.Diagnostics;
using System.IO;

using log4net;

using Axiom.Animating;

using Multiverse.Interface; // for UIScripting
using Multiverse.AssetRepository;
using Multiverse.Utility; // for TimeTool
using Multiverse.Lib.LogUtil;

namespace Multiverse.Base
{
    public delegate void WorldInitializedHandler(object sender, EventArgs e);
    public delegate void FrameStartedHandler(object sender, ScriptingFrameEventArgs e);
    public delegate void FrameEndedHandler(object sender, ScriptingFrameEventArgs e);

    public class ScriptingFrameEventArgs : EventArgs
    {
        protected float time;

        public ScriptingFrameEventArgs(float time)
            : base()
        {
            this.time = time;
        }

        public float TimeSinceLastFrame
        {
            get
            {
                return time;
            }
        }
    }

    public class ClientAPI
    {
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(typeof(ClientAPI));
        private static readonly log4net.ILog deprecatedLog = log4net.LogManager.GetLogger("ScriptDeprecated");

        protected static IGameWorld betaWorld;
        protected static bool initialized;
        protected static SortedList<long, List<object>> effectWorkQueue;
        protected static YieldEffectHandler yieldHandler;
        protected static Dictionary<AnimationState, AnimationStateInfo> playingAnimations;

        public static event WorldInitializedHandler WorldInitialized;
        public static event FrameStartedHandler FrameStarted;
        public static event FrameEndedHandler FrameEnded;
        public static event EventHandler WorldConnect;
        public static event EventHandler WorldDisconnect;

        static ClientAPI()
        {
            effectWorkQueue = new SortedList<long, List<object>>();
            playingAnimations = new Dictionary<AnimationState, AnimationStateInfo>();
        }

        public static void OnWorldInitialized() {
            WorldInitializedHandler handler = WorldInitialized;
            if (handler != null) {
                handler(null, new EventArgs());
            }
        }

        public static void OnWorldConnect() {
            if (WorldConnect != null)
                WorldConnect(null, new EventArgs());
        }
        public static void OnWorldDisconnect() {
            if (WorldDisconnect != null)
                WorldDisconnect(null, new EventArgs());
        }

        public static void OnFrameStarted(float time)
        {
            FrameStartedHandler handler = FrameStarted;
            if (handler != null)
            {
                handler(null, new ScriptingFrameEventArgs(time));
            }
        }

        public static void OnFrameEnded(float time)
        {
            FrameEndedHandler handler = FrameEnded;
            if (handler != null)
            {
                handler(null, new ScriptingFrameEventArgs(time));
            }
            //if (triggerWorldInitialized &&
                //betaWorld.WorldManager.SceneManager.CurrentViewport != null)
            if (triggerWorldInitialized) {
                OnWorldInitialized();
                triggerWorldInitialized = false;
            }
        }

        public static void InitAPI(IGameWorld gameWorld)
        {
            string scriptPath = null;
            // add to the path so that imports from ClientAPI will be found
            if (System.IO.Directory.Exists("..\\Scripts"))
            {
                scriptPath = "../Scripts/";
            }
            else
            {
                scriptPath = "../../Scripts/";
            }

            UiScripting.AddPath(scriptPath);

            log.InfoFormat("API Script Path: {0}", scriptPath);

            foreach (string dir in RepositoryClass.Instance.RepositoryDirectoryList) {
                string scriptRepository;
                scriptRepository = string.Format("{0}/Scripts/", dir);
                if (Directory.Exists(scriptRepository))
                {
                    UiScripting.AddPath(scriptRepository);
                    log.InfoFormat("World Script Path: {0}", scriptRepository);
                }
                scriptRepository = string.Format("{0}/IPCE/", dir);
                if (Directory.Exists(scriptRepository))
                {
                    UiScripting.AddPath(scriptRepository);
                    log.InfoFormat("World Script Path: {0}", scriptRepository);
                }
            }

            // Create a dictionary to initialize globals for the ClientAPI module
            Dictionary<string, object> globals = new Dictionary<string, object>();

            // Load the ClientAPI python code
            UiScripting.RunModule(scriptPath, "ClientAPI.py", "ClientAPI", true, globals);

            // Run world startup script
            if (!UiScripting.RunFile("Startup.py"))
                throw new PrettyClientException("bad_script.htm", "Unable to run startup scripts");

            yieldHandler = UiScripting.SetupDelegate<YieldEffectHandler>("return generator.next()", null);
            betaWorld = gameWorld;
            initialized = true;
        }

        public delegate int YieldEffectHandler(object generator);

        public static void QueueYieldEffect(object generator, int milliseconds)
        {
            long time = milliseconds + TimeTool.CurrentTime;
            List<object> timeList;

            if (effectWorkQueue.ContainsKey(time))
            {
                timeList = effectWorkQueue[time];
            }
            else
            {
                timeList = new List<object>();
                effectWorkQueue[time] = timeList;
            }
            timeList.Add(generator);
        }

        public static void ProcessYieldEffectQueue()
        {
            long currentTime = TimeTool.CurrentTime;

            // process any pending queue items
            while ((effectWorkQueue.Count > 0) && (effectWorkQueue.Keys[0] < currentTime))
            {
                // fetch the generator object
                List<object> generatorList = effectWorkQueue.Values[0];

                // remove it from the queue
                effectWorkQueue.RemoveAt(0);

                foreach (object generator in generatorList)
                {
                    int nextWait = 0;
                    bool done = false;

                    try
                    {
                        // call the python yield handler code with the generator object, which does
                        // the next slice of work on the effect.
                        nextWait = yieldHandler(generator);
                    }
                    catch (IronPython.Runtime.Exceptions.StopIterationException)
                    {
                        done = true;
                    }
                    catch (Exception ex)
                    {
                        string pystack = UiScripting.FormatException(ex);
                        LogUtil.ExceptionLog.ErrorFormat("Exception in yield effect handler.  Python Stack Trace: {0}\n.Full Stack Trace: {1}", pystack, ex);
                        log.Error("Cancelling effect execution.");
                        done = true;
                    }

                    if (!done)
                    {
                        // add it back to the queue at the next scheduled time offset
                        QueueYieldEffect(generator, nextWait);
                    }
                }
            }
        }

        public static AnimationStateInfo PlaySceneAnimation(AnimationState state, float speed, bool looping)
        {
            if (playingAnimations.ContainsKey(state))
            {
                log.ErrorFormat("Attempted to play an already playing animation: {0}", state.Name);
                return null;
            }

            AnimationStateInfo stateInfo = new AnimationStateInfo(state, speed, looping);

            playingAnimations[state] = stateInfo;
            state.Time = 0;

            return stateInfo;
        }

        public static void StopSceneAnimation(AnimationState state)
        {
            if (!playingAnimations.ContainsKey(state))
            {
                log.InfoFormat("Attempted to stop an animation that is not playing: {0}", state.Name);
                return;
            }

            playingAnimations.Remove(state);

            return;
        }

        public static void ProcessSceneAnimations(float timeSinceLastFrame)
        {
            if (playingAnimations.Count > 0)
            {
                List<AnimationState> removals = new List<AnimationState>();

                foreach (AnimationStateInfo info in playingAnimations.Values)
                {
                    float overflow = 0f;
                    if (info.AddTime(timeSinceLastFrame, out overflow))
                        removals.Add(info.State);
                }

                // remove any animations that have finished
                foreach (AnimationState state in removals)
                {
                    playingAnimations.Remove(state);
                }
            }
        }

        protected static bool triggerWorldInitialized;
        public static bool TriggerWorldInitialized
        {
            get
            {
                return triggerWorldInitialized;
            }
            set
            {
                triggerWorldInitialized = true;
            }
        }
        public static log4net.ILog Log {
            get { return log; }
        }

        protected static Dictionary<string, object> deprecatedCalls = new Dictionary<string, object>();

        public static void ScriptDeprecated(string version, string oldMethod, string newMethod)
        {
            StackTrace t = new StackTrace(true);
            StackFrame f = null;
            int i = 0;
            bool foundPython = false;

            // look for the first stack frame that appears to be in a python script
            for (; i < t.FrameCount; i++)
            {
                f = t.GetFrame(i);
                string filename = f.GetFileName();
                if ((filename != null) && filename.ToLowerInvariant().EndsWith(".py"))
                {
                    foundPython = true;
                    break;
                }
            }

            if (foundPython)
            {
                // generate a string that uniquely identifies a call to a deprecated interface
                //   from a particular line of script code.  We will use this string to avoid
                //   printing the same deprecated message over and over for the same line of code.
                string instanceID = oldMethod + f.GetFileName() + f.GetFileLineNumber().ToString();

                if (!deprecatedCalls.ContainsKey(instanceID))
                {
                    // mark the call instance ID in the 
                    deprecatedCalls[instanceID] = null;

                    deprecatedLog.WarnFormat("DEPRECATED:{0}: {1} in the Client API should no longer be used.  You should replace it with {2}", version, oldMethod, newMethod);
                    deprecatedLog.WarnFormat("  {0} is called:", oldMethod);

                    // continue from the index where the previous loop found a python file
                    for (; i < t.FrameCount; i++)
                    {
                        f = t.GetFrame(i);
                        string filename = f.GetFileName();
                        if ((filename != null) && filename.ToLowerInvariant().EndsWith(".py"))
                        {
                            deprecatedLog.WarnFormat("    at {0} in {1}: line {2}", f.GetMethod().Name, filename, f.GetFileLineNumber());
                        }
                    }
                }
            }
        }
    }
}
